File: Hazarath/2D CA-CFAR/2D_CA-CFAR.py
"""
Radar Target Generation and Detection -- small test-case variant
Based on: radar_target_generation_and_detection.py
(Python conversion of Radar_Target_Generation_and_Detection.m)
 
This variant is sized specifically to produce a 10 x 12 Range-Doppler Map
(RDM) so it can be used to validate a separately-written 2D CA-CFAR
implementation (e.g. in C) against this Python one:
 
    RDM size                : 10 (range bins) x 12 (doppler bins)
    Training cells (Tr, Td) : 2 x 2
    Guard cells    (Gr, Gd) : 2 x 2
 
Right after the Range-Doppler Map is generated, this script prints it out
as full-precision double values, formatted as a ready-to-paste C array
literal, so the exact same input matrix can be hardcoded into another
program for a side-by-side comparison.
"""
 
import numpy as np
 
# ---------------------------------------------------------------------------
# Radar Specifications
# ---------------------------------------------------------------------------
# Frequency of operation = 77GHz
# Max Range              = 200m
# Range Resolution        = 1 m
# Max Velocity            = 100 m/s
# speed of light          = 3e8
# ---------------------------------------------------------------------------
 
# ---------------------------------------------------------------------------
# User Defined Range and Velocity of target
# ---------------------------------------------------------------------------
# Define the target's initial position and velocity.
# Note: velocity remains constant.
v = -20
Range_init = 110
 
c = 3e8
r_res = 1
r_max = 200
 
# ---------------------------------------------------------------------------
# FMCW Waveform Generation
# ---------------------------------------------------------------------------
B_chirp = c / (2 * r_res)
T_chirp = 5.5 * (2 * r_max / c)
slope = B_chirp / T_chirp
 
# Operating carrier frequency of Radar
fc = 77e9  # carrier freq
 
# ---------------------------------------------------------------------------
# NOTE on sizing: Nr/2 rows x Nd columns is what ends up in the RDM (see the
# "RANGE DOPPLER RESPONSE" section below). To get a 10 x 12 RDM for CFAR
# testing purposes, Nr = 20 and Nd = 12. These are deliberately tiny
# compared to the original script's Nr=1024/Nd=128 -- this sizing is only
# meant to produce a small, easy-to-inspect matrix for algorithm
# validation, not a physically realistic range/Doppler resolution.
# ---------------------------------------------------------------------------
Nd = 12  # number of chirps -> RDM columns (doppler bins)
Nr = 20  # samples per chirp -> RDM has Nr/2 = 10 rows (range bins)
 
# Timestamp for running the displacement scenario for every sample on each
# chirp.
t = np.linspace(0, Nd * T_chirp, Nr * Nd)  # total time for samples
 
# ---------------------------------------------------------------------------
# Signal generation and Moving Target simulation (vectorized -- see the
# full-size script for the equivalent explicit per-sample for-loop).
# ---------------------------------------------------------------------------
r_t = Range_init + v * t
td = 2 * r_t / c
delay = t - td
 
Tx = np.cos(2 * np.pi * (fc * t + slope * (t ** 2) / 2))
Rx = np.cos(2 * np.pi * (fc * delay + slope * (delay ** 2) / 2))
Mix = Tx * Rx
 
# ---------------------------------------------------------------------------
# RANGE MEASUREMENT (1st FFT -- unrelated to CFAR, kept only for parity with
# the original script; skip straight to "RANGE DOPPLER RESPONSE" below if
# you only care about the CFAR input matrix).
# ---------------------------------------------------------------------------
Mix = Mix.reshape((Nr, Nd), order='F')
signal_fft = np.fft.fft(Mix, n=Nr, axis=0)
 
L = T_chirp * B_chirp
signal_fft = np.abs(signal_fft / L)
 
# NOTE: capped at Nr*Nd (total available samples) so this doesn't try to
# slice past the end of the array for a small Nr/Nd -- with the original
# script's Nr=1024/Nd=128 this cap never triggers and behavior is
# unchanged; it only matters for small test sizes like this one.
half = min(int(round(L / 2)) + 1, Nr * Nd)
signal_fft = signal_fft.flatten(order='F')[:half]
 
f = B_chirp * np.arange(half) / L
R = (c * T_chirp * f) / (2 * B_chirp)
 
# ---------------------------------------------------------------------------
# RANGE DOPPLER RESPONSE
# ---------------------------------------------------------------------------
Mix = Mix.reshape((Nr, Nd), order='F')
 
# 2D FFT using the FFT size for both dimensions.
sig_fft2 = np.fft.fft2(Mix, s=(Nr, Nd))
 
# Take just one side of the signal from the Range dimension.
sig_fft2 = sig_fft2[:Nr // 2, :Nd]
sig_fft2 = np.fft.fftshift(sig_fft2)
RDM = np.abs(sig_fft2)
RDM = 10 * np.log10(RDM)  # <-- this is "the Doppler map" (RDM), size 10 x 12
 
# ---------------------------------------------------------------------------
# Print the just-generated Range-Doppler Map as full double-precision
# values, formatted as a ready-to-paste C array literal. This is the exact
# matrix that the CFAR loop below operates on (before its own internal
# normalization step -- see the comment further down), so it's what you
# should feed into a separately-written CFAR implementation for a
# side-by-side comparison.
# ---------------------------------------------------------------------------
num_range_bins, num_doppler_bins = RDM.shape
 
# Also print the same matrix in a plain readable table, for a quick visual
# sanity check alongside the C literal above.
np.set_printoptions(precision=17, suppress=False, floatmode='maxprec_equal',
                     linewidth=200)
 
# ---------------------------------------------------------------------------
# CFAR implementation
# ---------------------------------------------------------------------------
# Number of Training Cells in both dimensions.
Tr = 2
Td = 3
 
# Number of Guard Cells in both dimensions around the Cell Under Test (CUT).
Gr = 2
Gd = 2
 
# Offset the threshold by an SNR value in dB.
offset = 1.2
 
 
def db2pow(db):
    """Convert dB to linear power (matches MATLAB's db2pow)."""
    return 10 ** (db / 10)
 
 
def pow2db(power):
    """Convert linear power to dB (matches MATLAB's pow2db)."""
    return 10 * np.log10(power)
 
 
# NOTE: this normalization happens *after* the RDM was already printed
# above, so the values you copied are the raw dB Range-Doppler Map, not
# this normalized version. If your own C implementation is meant to match
# this Python CFAR bit-for-bit, either feed it this same
# `RDM = RDM / max(RDM)` normalized matrix instead, or replicate this
# normalization step in your C code before running CFAR.
RDM = RDM / np.max(RDM)  # Normalizing

print(f"// Range-Doppler Map (RDM), {num_range_bins} x {num_doppler_bins}, "f"values in dB, as generated (double precision):")
print(f"double rdm[{num_range_bins}][{num_doppler_bins}] = {{")
for row in RDM:
    row_str = ", ".join(repr(float(value)) for value in row)
    print(f"    {{{row_str}}},")
print("};")
print()

for i in range(Tr + Gr, num_range_bins - (Tr + Gr)):
    for j in range(Td + Gd, num_doppler_bins - (Td + Gd)):
        noise_level = 0.0
        for p in range(i - (Tr + Gr), i + (Tr + Gr) + 1):
            for q in range(j - (Td + Gd), j + (Td + Gd) + 1):
                if abs(i - p) > Gr or abs(j - q) > Gd:
                    noise_level += db2pow(RDM[p, q])
        threshold = pow2db(
            noise_level
            / (2 * (Td + Gd + 1) * 2 * (Tr + Gr + 1) - (Gr * Gd) - 1)
        )
        threshold = threshold + offset
        CUT = RDM[i, j]
        if CUT < threshold:
            RDM[i, j] = 0
        else:
            RDM[i, j] = 1
 
RDM[(RDM != 0) & (RDM != 1)] = 0
 
print("CFAR output (0/1 detection map), for reference:")
print(RDM)

File: Hazarath/2D CA-CFAR/2D_CA-CFAR.c
// 2D CA-CFAR (Cell-Averaging Constant False Alarm Rate): adaptively flags a cell as a target when its value exceeds a noise threshold estimated by averaging the surrounding training cells, while excluding the guard cells closest to it.
#include<stdio.h>
#include<math.h>
#include<stdlib.h>
int main()
{
    // Input of Doppler 2D array cells, row and column-wise sizes of test and Guardian cells and offset.
    int input_x = 10,input_y = 12, Guardian_x = 2, Guardian_y = 2, Test_x = 2, Test_y = 3;
    double offset = 1.2, input[10][12] = {
    {0.21477920894998642, 0.5542752837260357, -0.0698928605142817, 0.49270772521570977, 0.7745915471662987, 0.40397069107175637, 0.4159834482266725, 0.5030152904717838, 0.5502494367567672, 0.5514778054668769, 0.542220799924258, 0.6101197486166718},
    {0.5908024698000002, 0.4607461046953283, 0.5643633942015163, 0.6549531011197849, 0.491725628590297, 0.38831411824404793, 0.5347171477907239, 0.7758433157252638, 0.308649842227188, 0.5351408865106796, 0.6547473274963175, 0.5465051370691714},
    {0.4392027753117354, 0.3254806417826386, 0.8031758834567246, 0.5095882114401862, 0.27941927200363254, 0.631058273372932, 0.45657214080005093, 0.6473794370306154, 0.3959325682402981, 0.0973001884802566, 0.002899055660089897, 0.6089976353542822},
    {0.020928815880501533, 0.8082826001951648, 0.7012047721196989, -0.006062465520964968, 0.5797593045170495, 0.2790103401571059, 0.46422664377958167, 0.6872084785819127, 0.3692362128659903, 0.7555001530959384, 0.6575975267907216, 0.420157705138798},
    {0.3970839637277771, 1.0, 0.6747442850785506, 0.5615516199857783, 0.583133178320232, 0.4124380149339856, 0.04087527857035471, 0.5593045378184145, 0.10382438905022086, 0.1511607801594206, 0.46564010131652006, 0.8422412739266505},
    {0.14331679007943274, 0.7254731583111648, 0.5068951547608074, 0.5627857759993787, 0.6838273172705014, 0.1919824584842269, 0.6964448536211438, 0.19198245848422693, 0.6838273172705012, 0.562785775999379, 0.5068951547608074, 0.7254731583111648},
    {0.39145123476161586, 0.6261305274176769, 0.567819066883081, 0.6001785320579135, 0.3856011902567108, 0.5708608620024291, 0.10345139064995093, 0.6062356634231407, 0.6594689966517595, 0.6078657018191677, 0.14848677380227374, 0.4782830201281335},
    {0.5798108362818903, 0.3424552461446489, 0.7749759517879828, 0.41760586700312124, 0.6421316511704441, 0.3695655010131712, 0.6905040665370221, 0.11198122186351957, 0.3986094539633821, 0.4523970891518089, 0.6509584965435949, 0.08454876663446892},
    {0.6387936229678775, 0.3570660112954653, 0.5871563683299293, 0.33659303292503123, 0.7260859410982176, -0.08116701712703125, 0.6697336333482686, 0.5217042726805496, 0.3622673718560428, 0.625287397769625, 0.6143013613531648, 0.7113100442005531},
    {0.7205871802391222, 0.09930345804791298, 0.3674682095222305, 0.27477610864330465, 0.5640634306881035, 0.4945170315105238, 0.6396985002518845, -0.03461632292757563, 0.40406942023147213, 0.7594189388458337, 0.3471187538700676, 0.2801204126047695},
    };
    
    //Start iterating through all the Test cells and check whether they are targets or not through 2D CA-CFAR algorithm
    for(int i=(Guardian_x+Test_x);i<(input_x)-(Guardian_x+Test_x);i++)
    {
        for(int j=(Guardian_y+Test_y);j<(input_y)-(Guardian_y+Test_y);j++)
        {
            //Implementing sliding-window accross all the cells in the calculated area. 
            //Iterating through the window and check if the value is lying outside the Guardian region or not
            double Noise_supression = 0.0;
            for(int p=i-(Guardian_x+Test_x);p<i+(Guardian_x+Test_x)+1;p++)
            {
                for(int q=j-(Guardian_y+Test_y);q<j+(Guardian_y+Test_y)+1;q++)
                {
                    //If cell is u=outside the Guardian region, consider it for noise supression calculation else do not consider.
                    if(abs(i-p)>Guardian_x || abs(j-q)>Guardian_y)
                    {
                        //Convert Decible to Linear power
                        Noise_supression += pow(10,(input[p][q]/10));
                    }
                }
            }
            //Implement Cell Averaging for refernece cells.
            double Threshold = (Noise_supression)/((2*(Guardian_x+Test_x+1)*2*(Guardian_y+Test_y+1))-(Guardian_x*Guardian_y)-1);
            //Convert linear power to Decible before addding the offset.
            Threshold = (10 * log10(Threshold));
            Threshold += offset;
            //If considered CUT is greater than threshold,CUT is considered to be a target.
            if(input[i][j]>=Threshold)
            {
                input[i][j]=1.0000;
            }
        }
    }
    //Targets are the one's having value of 1 as it's target cells and all the remaining one's are 0's.
    printf("The targets are:\n");
    for(int i=0;i<10;i++)
    {
        for(int j=0;j<12;j++)
        {
            if((int)input[i][j]==0||(int)input[i][j]==1)
            {
                printf("%d ",(int)input[i][j]);
            }
            else
            {
                printf("0 ");
            }
        }
        printf("\n");
    }
    return 0;
}

